### PR TITLE
Simple yk interpreters

### PR DESCRIPTION
**Pull request by ltratt**
In essence this PR adds two (very simple!) interpreters gradually adapt the base BF interpreter to be more-and-more yk friendly: though there is still much more that could be done! Both these compile but will panic at runtime (in yk's Rust code) because the HW mapper doesn't yet run.

It's a bit hard to see the diffs but:

```
  $ diff -u bf_base.c bf_simple_yk.c
  --- bf_base.c	Mon May 24 11:19:15 2021
  +++ bf_simple_yk.c	Mon May 24 12:01:44 2021
  @@ -5,13 +5,16 @@
   #include <stdlib.h>
   #include <sys/stat.h>
   #include <unistd.h>
  +#include <yk.h>

   #define CELLS_LEN 30000

  -void interp(char *prog, char *prog_end, char *cells, char *cells_end) {
  +void interp(char *prog, char *prog_end, char *cells, char *cells_end, YkLocation *yklocs) {
  +    YkMT *mt = yk_mt();
       char *instr = prog;
       char *cell = cells;
       while (instr < prog_end) {
  +        yk_control_point(mt, &yklocs[instr - prog]);
           switch (*instr) {
               case '>': {
                   if (cell++ == cells_end)
  @@ -93,5 +96,11 @@
       if (cells == NULL)
           err(1, "out of memory");

  -    interp(prog, prog + sb.st_size, cells, cells + CELLS_LEN);
  +    YkLocation *yklocs = malloc(sizeof(YkLocation) * CELLS_LEN);
  +    if (yklocs == NULL)
  +        err(1, "out of memory");
  +    for (YkLocation *ykloc = yklocs; ykloc < yklocs + CELLS_LEN; ykloc++)
  +        *ykloc = yk_location_new();
  +
  +    interp(prog, prog + sb.st_size, cells, cells + CELLS_LEN, yklocs);
   }
```

and:

```
  diff -u bf_simple_yk.c bf_simple2_yk.c
  --- bf_simple_yk.c	Mon May 24 12:01:44 2021
  +++ bf_simple2_yk.c	Mon May 24 12:01:40 2021
  @@ -9,7 +9,12 @@

   #define CELLS_LEN 30000

  +char *jmp_back(char *);
  +char *jmp_fwd(char *);
  +
   void interp(char *prog, char *prog_end, char *cells, char *cells_end, YkLocation *yklocs) {
  +    // FIXME: need to call yktrace_const on `prog` (or otherwise inform yk that
  +    // prog is immutable for the duration of a trace).
       YkMT *mt = yk_mt();
       char *instr = prog;
       char *cell = cells;
  @@ -46,37 +51,47 @@
               }
               case '[': {
                   if (*cell == 0) {
  -                    int count = 0;
  -                    while (true) {
  -                        instr++;
  -                        if (*instr == ']') {
  -                            if (count == 0)
  -                                break;
  -                            count--;
  -                        } else if (*instr == '[')
  -                            count++;
  -                    }
  +                    instr = jmp_fwd(instr);
                   }
                   break;
               }
               case ']': {
                   if (*cell != 0) {
  -                    int count = 0;
  -                    while (true) {
  -                        instr--;
  -                        if (*instr == '[') {
  -                            if (count == 0)
  -                                break;
  -                            count--;
  -                        } else if (*instr == ']')
  -                            count++;
  -                    }
  +                    instr = jmp_back(instr);
                   }
                   break;
               }
               default: break;
           }
           instr++;
  +    }
  +}
  +
  +// FIXME: needs to be given the yktrace_idempotent attribute
  +char *jmp_back(char *instr) {
  +    int count = 0;
  +    while (true) {
  +        instr--;
  +        if (*instr == '[') {
  +            if (count == 0)
  +                return instr;
  +            count--;
  +        } else if (*instr == ']')
  +            count++;
  +    }
  +}
  +
  +// FIXME: needs to be given the yktrace_idempotent attribute
  +char *jmp_fwd(char *instr) {
  +    int count = 0;
  +    while (true) {
  +        instr++;
  +        if (*instr == ']') {
  +            if (count == 0)
  +                return instr;
  +            count--;
  +        } else if (*instr == '[')
  +            count++;
       }
   }
  $
```

### CODE REVIEW

**Comment by ltratt** [comment]
I need to rebase this against master. OK for me to do so?

**Comment by vext01** [comment]
Please.

**Comment by ltratt** [comment]
Squashed. I've also done a fair bit of work on `.buildbot.sh` which I'm bound not to have got right first time around...

bors try

File: .buildbot.sh [hunk 1]
```diff
@@ -2,19 +2,42 @@
...
 sh rustup.sh --default-host x86_64-unknown-linux-gnu \
-    --default-toolchain stable \
+    --default-toolchain nightly \
     --no-modify-path \
     --profile minimal \
     -y
 export PATH=`pwd`/.cargo/bin/:$PATH
 
-rustup toolchain install stable
+rustup toolchain install nightly
```
Comment (vext01) [inline .buildbot.sh]: Is this needed? Above the default is already nightly?

Comment (ltratt) [reply .buildbot.sh]: I don't know -- I'm not a rustup person!

File: Makefile [hunk 2]
```diff
@@ -1,10 +1,22 @@
...
 
-bf: bf.o
-	${CC} ${LDFLAGS} -o bf bf.o
+bf_base: bf_base.o
+	clang -o bf_base bf_base.o
 
-bf.o: bf.c
-	${CC} ${CFLAGS} -Wall -c bf.c
+bf_base.o: bf_base.c
+	clang -Wall -O3 -c bf_base.c
```
Comment (vext01) [inline Makefile]: Some targets have `-O3 ${CFLAGS}` and others only `-O3`. Is that intentional?

Comment (ltratt) [reply Makefile]: Fixed in a1d505b.

**Comment by ltratt** [comment]
bors try

File: .buildbot.sh (hunk 1 above)
Comment (vext01) [reply .buildbot.sh]: It looks to me like you can remove this step.

Comment (ltratt) [reply .buildbot.sh]: Fixed in 1e56e63.

**Comment by vext01** [comment]
Please squash.

**Comment by ltratt** [comment]
Squashed.

**Pull request merged**